core/arch/arm/plat-imx/drivers/imx_priblob.c

```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2020 Pengutronix, Rouven Czerwinski <entwicklung@pengutronix.de>
+ */
+#include <config.h>
+#include <initcall.h>
+#include <io.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+
+#define SCFGR		0x000C
+#define	PRIBLOB_11	0x3
+#define PRIBLOB_MASK	GENMASK_32(1, 0)
+
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, CAAM_BASE, CORE_MMU_PGDIR_SIZE);
+
+static TEE_Result lock_priblob(void)
```

Comment:
Why implementing this file.
1. You are redefining SCFGR which is already defined in the caam/hal/imx_[6_7/8m]/registers/ctrl_regs.h
2. This function must be part of the caam/hal/common/hal_ctrl.c file
3. We have to ensure that this function is callled before leaving the secure environment (when we know that all drivers needing to create a secure key derived from master HW key are done)

In the commit description you are explaining that this function is used to increment PRIBLOB but here is not not incremental this is a set to 0x3. To increment, you have to read, add one and write the PRIBLOB.

Reply:
In https://github.com/OP-TEE/optee_os/pull/3608#discussion_r381135183 you said
> Hence, setting the PRIBLOB to 0x3 must remain a platform specific workaround done before switching to normal OS boot. It can not be part of CAAM driver at all.

Which is now implemented as a `plat-imx` specific workaround. I interpreted this as a request to remove this part from the CAAM driver completely. Please describe in detail what you had in mind for this change.

Reply:
Problem with the proposed implementation is that it's add a new driver file in plat-imx directory and redefining stuff already done, like the base address of the register and the register offset. Work is duplicated and so if we have to fix/maintain we have to do it at different places.
Keeping function in the already existing caam driver (in the appropriate file) remove risk of bug and confusion during fix/maintenance/porting.
You can keep this operation as a workaround with a CFG_xxx flag in the CAAM driver itself making it in generic way (I mind transparently regardless device and if workaround must be applied).

The workaround concerns the CAAM driver hence it must be in CAAM driver.

HAL directory is here to maintain difference between i.MX and its help so much when porting driver between devices. Keeping the same logic even for workaround help also during porting to other devices. In other words, common driver logic/implementation doesn't have to be touched to add/fix specific devices, only hal has to be changed.